services/src/contexts/migrations/current_schema.sql

```diff
@@ -1168,60 +1168,76 @@ CREATE UNIQUE INDEX ON permissions (
     ml_model_id
 );
 
+CREATE UNIQUE INDEX ON permissions (
+    role_id,
+    permission,
+    provider_id
+);
+
 CREATE VIEW user_permitted_datasets
 AS
 SELECT
     r.user_id,
     p.dataset_id,
-    p.permission
+    MAX(p.permission) AS max_permission
 FROM user_roles AS r
-INNER JOIN permissions AS p ON (
-    r.role_id = p.role_id AND p.dataset_id IS NOT NULL
-);
+INNER JOIN permissions AS p
+    ON (
+        r.role_id = p.role_id AND p.dataset_id IS NOT NULL
+    )
+GROUP BY r.user_id, p.dataset_id;
 
 CREATE VIEW user_permitted_projects
 AS
 SELECT
     r.user_id,
     p.project_id,
-    p.permission
+    MAX(p.permission) AS max_permission
 FROM user_roles AS r
-INNER JOIN permissions AS p ON (
-    r.role_id = p.role_id AND p.project_id IS NOT NULL
-);
+INNER JOIN permissions AS p
+    ON (
+        r.role_id = p.role_id AND p.project_id IS NOT NULL
+    )
+GROUP BY r.user_id, p.project_id;
 
 CREATE VIEW user_permitted_layer_collections
 AS
 SELECT
     r.user_id,
     p.layer_collection_id,
-    p.permission
+    MAX(p.permission) AS max_permission
 FROM user_roles AS r
-INNER JOIN permissions AS p ON (
-    r.role_id = p.role_id AND p.layer_collection_id IS NOT NULL
-);
+INNER JOIN permissions AS p
+    ON (
+        r.role_id = p.role_id AND p.layer_collection_id IS NOT NULL
+    )
+GROUP BY r.user_id, p.layer_collection_id;
 
 CREATE VIEW user_permitted_layers
 AS
 SELECT
     r.user_id,
     p.layer_id,
-    p.permission
+    MAX(p.permission) AS max_permission
```

Comment:
This works because enums are ints and read=0, owner=1?

Reply:
Yes, exactly